File: backend/services/doctorFormsStore.js
const SpecialistMatchRequest = require('../models/SpecialistMatchRequest');
const ProviderApplication = require('../models/ProviderApplication');

function isValidEmail(email) {
  return /^[^@\s]+@[^@\s]+\.[^@\s]+$/.test(email.trim());
}

async function saveSpecialistMatchRequest(body) {
  const submitterName = (body.submitterName || body.name || '').trim();
  const submitterEmail = (body.submitterEmail || body.email || '').trim();
  const doctorName = (body.doctorName || '').trim();
  const specialty = (body.specialty || '').trim();
  const location = (body.location || '').trim();
  const details = (body.details || body.description || '').trim();

  if (!doctorName) throw new Error('Doctor name is required');
  if (!specialty) throw new Error('Specialty is required');
  if (!location) throw new Error('Location is required');
  if (submitterEmail && !isValidEmail(submitterEmail)) {
    throw new Error('Please provide a valid email address');
  }

  const record = await SpecialistMatchRequest.create({
    submitterName,
    submitterEmail,
    doctorName,
    specialty,
    location,
    details,
  });

  return {
    success: true,
    message: 'Thank you! We have received your specialist suggestion.',
    requestId: String(record._id),
  };
}

async function saveProviderApplication(body) {
  const name = (body.name || '').trim();
  const specialty = (body.specialty || '').trim();
  const description = (body.description || '').trim();
  const email = (body.email || '').trim();
  const phone = (body.phone || '').trim();

  if (!name) throw new Error('Name is required');
  if (!specialty) throw new Error('Specialty is required');
  if (!description) throw new Error('Description is required');
  if (email && !isValidEmail(email)) {
    throw new Error('Please provide a valid email address');
  }

  const record = await ProviderApplication.create({
    name,
    specialty,
    description,
    email,
    phone,
  });

  return {
    success: true,
    message: 'Thank you for your interest in joining OvaCare. We will be in touch soon.',
    requestId: String(record._id),
  };
}

module.exports = {
  saveSpecialistMatchRequest,
  saveProviderApplication,
};

File: backend/services/doctor_forms_store.py
from __future__ import annotations

import re
from datetime import datetime, timezone
from typing import Any, Dict

from services.mongodb_client import get_collection


def _is_valid_email(email: str) -> bool:
    return bool(re.match(r"^[^@\s]+@[^@\s]+\.[^@\s]+$", email.strip()))


def save_specialist_match_request(
    *,
    submitter_name: str,
    submitter_email: str,
    doctor_name: str,
    specialty: str,
    location: str,
    details: str,
) -> Dict[str, Any]:
    submitter_name = submitter_name.strip()
    submitter_email = submitter_email.strip()
    doctor_name = doctor_name.strip()
    specialty = specialty.strip()
    location = location.strip()
    details = details.strip()

    if not doctor_name:
        raise ValueError("Doctor name is required")
    if not specialty:
        raise ValueError("Specialty is required")
    if not location:
        raise ValueError("Location is required")
    if submitter_email and not _is_valid_email(submitter_email):
        raise ValueError("Please provide a valid email address")

    now = datetime.now(timezone.utc)
    record = {
        "submitterName": submitter_name,
        "submitterEmail": submitter_email,
        "doctorName": doctor_name,
        "specialty": specialty,
        "location": location,
        "details": details,
        "status": "pending",
        "createdAt": now,
        "updatedAt": now,
    }
    result = get_collection("specialistmatchrequests").insert_one(record)
    return {"success": True, "requestId": str(result.inserted_id)}


def save_provider_application(
    *,
    name: str,
    specialty: str,
    description: str,
    email: str = "",
    phone: str = "",
) -> Dict[str, Any]:
    name = name.strip()
    specialty = specialty.strip()
    description = description.strip()
    email = email.strip()
    phone = phone.strip()

    if not name:
        raise ValueError("Name is required")
    if not specialty:
        raise ValueError("Specialty is required")
    if not description:
        raise ValueError("Description is required")
    if email and not _is_valid_email(email):
        raise ValueError("Please provide a valid email address")

    now = datetime.now(timezone.utc)
    record = {
        "name": name,
        "specialty": specialty,
        "description": description,
        "email": email,
        "phone": phone,
        "status": "pending",
        "createdAt": now,
        "updatedAt": now,
    }
    result = get_collection("providerapplications").insert_one(record)
    return {"success": True, "requestId": str(result.inserted_id)}
